File: Gemi3/pyCAVI/_update_y.py
import numpy as np
import sys
from Gemi3.pyCAVI.util import find_s_keys,find_r_keys,gene_to_guides_fast

def y_loop(self):
    if self.verbose:
        print("Updating y values.",file=sys.stdout)
    for gl in self.hash_y.keys():
        indices=collect_all_keys(self,gl)
        numerator_sum=np.sum(np.asarray(list(map(lambda x: y_numerator(self,x),indices))))
        denominator_sum=np.sum(np.asarray(list(map(lambda x: y_denominator(self,x),indices))))
        new_ygl=((self.prio_tau_y*self.prio_mu_y)+numerator_sum)/(self.prio_tau_y+denominator_sum)
        if self.debug: print(f"Old ygl {self.y[self.hash_y[gl]]}.", sys.stdout)
        self.y[self.hash_y[gl]] = new_ygl
        if self.debug: print(f"New ygl {self.y[self.hash_y[gl]]}.", sys.stdout)


def yy_loop(self):
    if self.verbose:
        print("Updating yy values.",file=sys.stdout)
    for gl in self.hash_y.keys():
        indices=collect_all_keys(self,gl)
        numerator_sum=np.sum(np.asarray(list(map(lambda x: y_numerator(self,x),indices))))
        denominator_sum=np.sum(np.asarray(list(map(lambda x: y_denominator(self,x),indices))))
        new_yygl=((self.prio_tau_yy*self.prio_mu_yy)+numerator_sum)/(self.prio_tau_yy+denominator_sum)
        if self.debug: print(f"Old yygl {self.yy[self.hash_y[gl]]}.", sys.stdout)
        self.yy[self.hash_y[gl]] = new_yygl
        if self.debug: print(f"New yygl {self.yy[self.hash_y[gl]]}.", sys.stdout)

def y_numerator(self,key_dict):
    numerator=(self.tau[self.hash_t[key_dict["t"]]]*self.x[self.hash_x[key_dict["gi"]]])*\
    (self.D[self.hash_d[key_dict["t"]]]-\
    (self.x[self.hash_x[key_dict["hj"]]]*self.y[self.hash_y[key_dict["hl"]]])-\
    (self.x[self.hash_x[key_dict["fk"]]]*self.y[self.hash_y[key_dict["fl"]]])-\
    (self.r[self.hash_r[key_dict["r"]]]*self.s[self.hash_s[key_dict["s"]]]))
    return numerator

def y_denominator(self,key_dict):
    denominator = self.tau[self.hash_t[key_dict["t"]]] * self.xx[self.hash_x[key_dict["gi"]]]
    return denominator

def collect_all_keys(self,gl_extern):
    #only get keys from tau and split them accordingly
    gene,sample=gl_extern.split("_")
    gi_extern=gene_to_guides_fast(self,gene)
    tau_keys = [x for x in self.hash_t.keys() if any(list(map(lambda d: d in x,gi_extern)))]
    out=[]
    for k in tau_keys:
        gi,hj,fk,l = k.split("_")
        gl = self.guide_to_gene(gi)
        hl = self.guide_to_gene(hj)
        fl = self.guide_to_gene(fk)
        s = find_s_keys(self,gl,hl,fl,l)
        r = find_r_keys(self,gi,hj,fk)
        out.append({"t":k,"gi":gi,"hj":hj,"fk":fk,"l":l,
                    "gl":"_".join([gl,l]),
                    "hl":"_".join([hl,l]),
                    "fl":"_".join([hl,l]),
                    "r":r,
                    "s":s})
    return out

File: src/CAVI/update_y.rs
use super::CAVI;
use std::collections::HashMap;
use itertools::Itertools;
use rayon::prelude::*;

impl<'a> CAVI<'a>{
    pub fn update_y(&self) -> Result<(),Box<dyn std::error::Error>>{
        if self.verbose {
            println! {"Updating y values"}
        }
        for gl in self.lookup.read().unwrap().get("y").unwrap().keys(){
            let all_numerators :Vec<f64> = self.lookup.read().unwrap().get("y").unwrap().get(gl).unwrap().par_iter().map(|element| y_numerator(self,element)).collect();
            let all_denominators :Vec<f64> = self.lookup.read().unwrap().get("y").unwrap().get(gl).unwrap().par_iter().map(|element| y_denominator(self,element)).collect();
            let new_ygl : f64 = ((self.prio_tau_y * self.prio_mu_y) + all_numerators.par_iter().sum::<f64>())/(self.prio_tau_y+all_denominators.par_iter().sum::<f64>());
            if self.debug {println!("Old ygl {}.",*self.all_y.read().unwrap().get(*self.hash_y.get(*gl).unwrap()).unwrap())}
            {*self.all_y.write().unwrap().get_mut(*self.hash_y.get(*gl).unwrap()).unwrap()= new_ygl}
            if self.debug {println!("New ygl {}.",*self.all_y.read().unwrap().get(*self.hash_y.get(*gl).unwrap()).unwrap())}
        }
        return Ok(());
    }

    pub fn update_yy(&self) -> Result<(),Box<dyn std::error::Error>>{
        if self.verbose{
            println!{"Updating yy values"}
        }
        for gl in self.lookup.read().unwrap().get("y").unwrap().keys(){
            let all_numerators :Vec<f64> = self.lookup.read().unwrap().get("y").unwrap().get(gl).unwrap().par_iter().map(|element| y_numerator(self,element)).collect();
            let all_denominators :Vec<f64> = self.lookup.read().unwrap().get("y").unwrap().get(gl).unwrap().par_iter().map(|element| y_denominator(self,element)).collect();
            let new_yygl : f64 = ((self.prio_tau_yy * self.prio_mu_yy) + all_numerators.par_iter().sum::<f64>())/(self.prio_tau_yy+all_denominators.par_iter().sum::<f64>());
            if self.debug {println!("Old yygl {}.",*self.all_yy.read().unwrap().get(*self.hash_y.get(*gl).unwrap()).unwrap())}
            {*self.all_yy.write().unwrap().get_mut(*self.hash_y.get(*gl).unwrap()).unwrap()= new_yygl}
            if self.debug {println!("New yygl {}.",*self.all_yy.read().unwrap().get(*self.hash_y.get(*gl).unwrap()).unwrap())}
        }
        return Ok(());
    }

}

fn y_numerator(obj : &CAVI,key_dict : &HashMap<&str,&str>) -> f64 {
    let numerator : f64 = (*obj.all_tau.read().unwrap().get(*obj.hash_td.get(*key_dict.get("t").unwrap()).unwrap()).unwrap() *
        *obj.all_x.read().unwrap().get(*obj.hash_x.get(*key_dict.get("gi").unwrap()).unwrap()).unwrap()) *
        (*obj.all_D.read().unwrap().get(*obj.hash_td.get(*key_dict.get("t").unwrap()).unwrap()).unwrap() -
        *obj.all_x.read().unwrap().get(*obj.hash_x.get(*key_dict.get("hj").unwrap()).unwrap()).unwrap() * *obj.all_y.read().unwrap().get(*obj.hash_y.get(*key_dict.get("hl").unwrap()).unwrap()).unwrap() -
        *obj.all_x.read().unwrap().get(*obj.hash_x.get(*key_dict.get("fk").unwrap()).unwrap()).unwrap() * *obj.all_y.read().unwrap().get(*obj.hash_y.get(*key_dict.get("fl").unwrap()).unwrap()).unwrap() -
        *obj.all_r.read().unwrap().get(*obj.hash_r.get(*key_dict.get("r").unwrap()).unwrap()).unwrap() * *obj.all_s.read().unwrap().get(*obj.hash_s.get(*key_dict.get("s").unwrap()).unwrap()).unwrap());
    return numerator;
    }


fn y_denominator(obj : &CAVI,key_dict : &HashMap<&str,&str>) -> f64 {
    let denominator : f64 = *obj.all_tau.read().unwrap().get(*obj.hash_td.get(*key_dict.get("t").unwrap()).unwrap()).unwrap() *
        *obj.all_xx.read().unwrap().get(*obj.hash_x.get(*key_dict.get("gi").unwrap()).unwrap()).unwrap();
    return denominator;
    }
